pkg/lib/aws/servicequotas.go

```diff
@@ -103,3 +110,156 @@ func (c *Client) VerifyInstanceQuota(instanceType string, requiredOnDemandInstan
+			}
+			for _, quota := range page.Quotas {
+				if quota == nil || quota.QuotaCode == nil || quota.Value == nil {
+					continue
+				}
+				if _, ok := vpcQuotaCodeToValueMap[*quota.QuotaCode]; ok {
+					vpcQuotaCodeToValueMap[*quota.QuotaCode] = int(*quota.Value)
+				}
+			}
+			return true
+		},
+	)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get EIP in use
+	elasticIPsInUse, err := c.ListElasticIPs()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get IGW in use
+	internetGatewaysInUse, err := c.ListInternetGateways()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// get NAT GW in use per selected AZ
+	gateways, err := c.DescribeNATGateways()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	subnets, err := c.DescribeSubnets()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	gatewaysSubnetIDs := SubnetIDsFromNATGateways(gateways)
+	azToGatewaysInUse := make(map[string]int)
+	for _, gatewaySubnetID := range gatewaysSubnetIDs {
+		for _, subnet := range subnets {
+			if subnet.SubnetId == nil || subnet.AvailabilityZone == nil {
+				continue
+			}
+			if !availabilityZones.Has(*subnet.AvailabilityZone) {
+				continue
+			}
+			if *subnet.SubnetId == gatewaySubnetID {
+				if _, ok := azToGatewaysInUse[*subnet.AvailabilityZone]; !ok {
+					azToGatewaysInUse[*subnet.AvailabilityZone] = 1
+				} else {
+					azToGatewaysInUse[*subnet.AvailabilityZone]++
+				}
+			}
+		}
+	}
+
+	// get VPC IDs
+	vpcs, err := c.DescribeVpcs()
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// check NAT GW quota
+	numOfExhaustedNATGatewayAZs := 0
+	greatestNATGatewayQuotaDeficit := 0
+	azsWithQuotaDeficit := []string{}
+	for az, numActiveGatewaysOnAZ := range azToGatewaysInUse {
+		azDeficit := vpcQuotaCodeToValueMap[_natGatewayQuotaCode] - numActiveGatewaysOnAZ - requiredNATGatewaysPerAZ
+		if azDeficit < 0 {
+			numOfExhaustedNATGatewayAZs++
+			azsWithQuotaDeficit = append(azsWithQuotaDeficit, az)
+			if -azDeficit > greatestNATGatewayQuotaDeficit {
+				greatestNATGatewayQuotaDeficit = -azDeficit
+			}
+		}
+	}
+	if highlyAvailableNATGateway && numOfExhaustedNATGatewayAZs > 0 {
+		return ErrorNATGatewayLimitExceeded(vpcQuotaCodeToValueMap[_natGatewayQuotaCode], greatestNATGatewayQuotaDeficit, azsWithQuotaDeficit, c.Region)
+	} else if !highlyAvailableNATGateway && numOfExhaustedNATGatewayAZs == len(availabilityZones) {
+		return ErrorNATGatewayLimitExceededInAllAZs(vpcQuotaCodeToValueMap[_natGatewayQuotaCode], greatestNATGatewayQuotaDeficit, c.Region)
+	}
+
+	// check EIP quota
+	var requiredElasticIPs int
+	if requiredNATGatewaysPerAZ > 0 {
+		if highlyAvailableNATGateway {
+			requiredElasticIPs = len(availabilityZones)
+		} else {
+			requiredElasticIPs = 1
+		}
+	}
+	if ec2QuotaCodeToValueMap[_elasticIPsQuotaCode]-len(elasticIPsInUse)-requiredElasticIPs < 0 {
+		additionalQuotaRequired := -ec2QuotaCodeToValueMap[_elasticIPsQuotaCode] + len(elasticIPsInUse) + requiredElasticIPs
```

Comment:
nit: it might be easier to read if we put the - at the end, e.g. `additionalQuotaRequired := len(elasticIPsInUse) + requiredElasticIPs - ec2QuotaCodeToValueMap[_elasticIPsQuotaCode]`.

If you agree, this would also apply to the next two